NeosModLoader/Utility/GenericTypeMethodsInvoker.cs

```diff
@@ -0,0 +1,166 @@
+		{
+			return (TReturn)InvokeInternal(method, instanceTypes, instance, methodType, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type[] instanceTypes, object instance, Type[] methodTypes, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceTypes, instance, methodTypes, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type instanceType, object instance, Type[] methodTypes, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceType, instance, methodTypes, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type[] instanceTypes, object instance, Type methodType, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceTypes, instance, methodType, parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type instanceType, object instance, Type methodType, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceType, instance, methodType, parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type instanceType, object instance, Type[] methodTypes, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceType, instance, methodTypes, parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type instanceType, object instance, Type methodType, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceType, instance, methodType, parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type instanceType, object instance, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceType, instance, new TypeDefinition(), parameters);
+		}
+
+		public TReturn Invoke<TReturn>(MethodInfo method, Type instanceType, params object[] parameters)
+		{
+			return (TReturn)InvokeInternal(method, instanceType, null, new TypeDefinition(), parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type[] instanceTypes, object instance, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceTypes, instance, new TypeDefinition(), parameters);
+		}
+
+		public object Invoke(MethodInfo method, Type[] instanceTypes, params object[] parameters)
+		{
+			return InvokeInternal(method, instanceTypes, null, new TypeDefinition(), parameters);
+		}
+
+		private static MethodInfo GetGenericMethodOfConcreteTypeDefault(MethodInfo needleMethod, Type concreteType)
+		{
+			NeosMod.Debug($"Looking for: {needleMethod.ReturnType.Name} {needleMethod.Name}({string.Join(", ", needleMethod.GetParameters().Select(p => $"{p.ParameterType.Name} {p.Name}"))})");
+
+			return concreteType.GetMethods(AccessTools.all)
+				.Single(hayMethod =>
+				{
+					NeosMod.Debug($"Testing: {hayMethod.ReturnType.Name} {hayMethod.Name}({string.Join(", ", hayMethod.GetParameters().Select(p => $"{p.ParameterType.Name} {p.Name}"))})");
```

Comment:
The debug logs should probs use the lambda function variant to avoid running the code if debug mode is not enabled...?

Reply:
I'm not sure if it makes sense to keep the debug prints around anyways

Reply:
I'm fine with keeping debug logs around if they:
- aren't incredibly spammy
- would be helpful if something goes wrong in NML

However, you should be using Logger.DebugInternal or DebugFuncInternal for NML-generated logs. The NeosMod.Debug function does a bunch of extra baloney to figure out what mod called it that NML internally doesn't need.

Reply:
> However, you should be using Logger.DebugInternal or DebugFuncInternal for NML-generated logs. The NeosMod.Debug function does a bunch of extra baloney to figure out what mod called it that NML internally doesn't need.

Ah yes, I didn't think about that when quickly modifying it from the version I wrote inside a mod project :'D